### PR TITLE
Add support for label permissions

### PR DESCRIPTION
**Pull request by stnguyen90**
A user can be given a particular label and access to resources can be restricted to only users with that label.

### CODE REVIEW

File: src/Database/Validator/Alphanumeric.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
+<?php
+
+namespace Utopia\Database\Validator;
+
+use Utopia\Validator\Text;
+
+class Alphanumeric extends Text
+{
+    /**
```
Comment (eldadfux) [inline src/Database/Validator/Alphanumeric.php:9]: Can we reuse some other validators from the framework or maybe the key validator for consistent logic?

File: src/Database/Validator/Label.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+class Label extends Alphanumeric
+{
+    /**
+     * Label constructor.
+     *
+     * Validate text is a valid label
+     *
+     */
+    public function __construct()
```
Comment (eldadfux) [inline src/Database/Validator/Label.php:13]: If we do have a dedicated label validator, not sure the `Alphanumeric` will be reused any time soon. I would vote against pre optimization.

File: src/Database/Validator/Alphanumeric.php (hunk 1 above)
Comment (stnguyen90) [reply src/Database/Validator/Alphanumeric.php:9]: I can re-use the `Text` validator, but the error message is very messy:

```json
{
    "message": "Invalid labels: Value must a valid array and Value must be a valid string and at least 1 chars and no longer than 36 chars and only consist of '0, 1, 2, 3, 4, 5, 6, 7, 8, 9, A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z, a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z' chars",
    "code": 400,
    "type": "general_argument_invalid",
    "version": "dev"
}
```

`Key` validator would require additional logic to check for special characters that it's not really worth it:

```php
    public function isValid($value): bool
    {
        if (!\is_string($value)) {
            return false;
        }

        /** NOT NEEDED

        // no leading special characters
        $leading = \mb_substr($value, 0, 1);
        if ($leading === '_' || $leading === '.' || $leading === '-') {
            return false;
        }

        $isInternal = $leading === '$';


        if ($isInternal && !$this->allowInternal) {
            return false;
        }

        if ($isInternal) {
            $allowList = [ '$id', '$createdAt', '$updatedAt' ];

            // If exact match, no need for any further checks
            return \in_array($value, $allowList);
        }
        
        **/
        
        // Valid chars: A-Z, a-z, 0-9, underscore, hyphen, period
        if (\preg_match('/[^A-Za-z0-9\_\-\.]/', $value)) {        // this regex can be simpler too
            return false;
        }

        if (\mb_strlen($value) > 36) {
            return false;
        }

        return true;
    }
```

Comment (stnguyen90) [reply src/Database/Validator/Alphanumeric.php:9]: Here are the two approaches:

```php
class Label extends Key
{
    protected string $message = 'Value must be a valid string between 1 and 36 chars containing only alphanumeric chars';

    /**
     * Is valid.
     *
     * Returns true if valid or false if not.
     *
     * @param $value
     *
     * @return bool
     */
    public function isValid($value): bool
    {
        if (!parent::isValid($value)) {
            return false;
        }

        // Valid chars: A-Z, a-z, 0-9
        if (\preg_match('/[^A-Za-z0-9]/', $value)) {
            return false;
        }
        
        return true;
    }
}
```

```php
class Label extends Text
{
    /**
     * Label constructor.
     */
    public function __construct()
    {
        parent::__construct(36, 1, \array_merge(Text::ALPHABET_UPPER, Text::ALPHABET_LOWER, Text::NUMBERS));
    }

    /**
     * Get Description
     *
     * Returns validator description
     *
     * @return string
     */
    public function getDescription(): string
    {
        return 'Value must be a valid string between 1 and 36 chars containing only alphanumeric chars';
    }
}
```

Comment (eldadfux) [reply src/Database/Validator/Alphanumeric.php]: I'm happy with either of them. If we need the framework as a dependency just for the Text validator, I would lean towards the first solution.

**Review by eldadfux — APPROVED** [review]

**Pull request merged**